### PR TITLE
Fix connection leak in the celery_worker pytest fixture

### LINKED ISSUE #6382: celery_worker fixture doesn't close redis connections
# Checklist

* [ ] I have verified that the issue exists against the `master` branch of Celery.
* [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
* [ ] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
* [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
* [ ] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
* [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
* [ ] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information
* [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
* [ ] I have verified that the issue exists against the `master` branch of Celery.
* [x] I have included the contents of ``pip freeze`` in the issue.
* [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

* [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
* [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
* [x] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
* [ ] I have tried reproducing the issue on more than one operating system.
* [ ] I have tried reproducing the issue on more than one workers pool.
* [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
* [x] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues
* None

#### Possible Duplicates
* None

## Environment & Settings

**Celery version**: celery==5.0.0

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:4.4.5 (cliffs) kombu:4.6.10 py:3.8.3
            billiard:3.6.3.0 py-amqp:2.6.0
platform -> system:Darwin arch:64bit
            kernel version:19.6.0 imp:CPython
loader   -> celery.loaders.default.Loader
settings -> transport:amqp results:disabled
```

</p>
</details>

# Steps to Reproduce
Run `py.test` on:

```python
"""
Redis connection tests
"""

import time


def test_one(celery_worker):

    pass


def test_two():

    while True:
        print("sleep...")
        time.sleep(1)

```

and check open connections with
`watch -n 0.1 'lsof -a -d 0-999 | grep ${REDIS_PORT}'`
`celery_worker` fixture of `test_one` opens connections to redis, but they are never closed.
This becomes a serious problems once you have hundreds of tests - we are running into max open files descriptors error due to this.

## Required Dependencies

* **Minimal Python Version**: 3.8
* **Minimal Celery Version**: 4.4.5 and 5.0
* **Minimal Kombu Version**: 5.0.2
* **Minimal Broker Version**: redis==3.5.3
* **Minimal Result Backend Version**: redis:6.0.5-buster
* **Minimal OS and/or Kernel Version**: OSX 0.15.6
* **Minimal Broker Client Version**: redis==3.5.3
* **Minimal Result Backend Client Version**: redis==3.5.3

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>
absl-py==0.9.0
adal==1.2.4
alembic==1.4.2
altair==4.1.0
amqp==2.6.0
antlr4-python3-runtime==4.7.2
applicationinsights==0.11.9
appnope==0.1.0
argcomplete==1.11.1
astor==0.8.1
astroid==2.4.2
astunparse==1.6.3
attrs==19.3.0
azure-batch==9.0.0
azure-cli==2.8.0
azure-cli-command-modules-nspkg==2.0.3
azure-cli-core==2.8.0
azure-cli-nspkg==3.0.4
azure-cli-telemetry==1.0.4
azure-common==1.1.25
azure-core==1.7.0
azure-cosmos==3.1.2
azure-datalake-store==0.0.48
azure-functions-devops-build==0.0.22
azure-graphrbac==0.60.0
azure-keyvault==1.1.0
azure-loganalytics==0.1.0
azure-mgmt-advisor==2.0.1
azure-mgmt-apimanagement==0.1.0
azure-mgmt-appconfiguration==0.4.0
azure-mgmt-applicationinsights==0.1.1
azure-mgmt-authorization==0.52.0
azure-mgmt-batch==9.0.0
azure-mgmt-batchai==2.0.0
azure-mgmt-billing==0.2.0
azure-mgmt-botservice==0.2.0
azure-mgmt-cdn==4.1.0rc1
azure-mgmt-cognitiveservices==6.2.0
azure-mgmt-compute==12.1.0
azure-mgmt-consumption==2.0.0
azure-mgmt-containerinstance==1.5.0
azure-mgmt-containerregistry==3.0.0rc13
azure-mgmt-containerservice==9.0.1
azure-mgmt-core==1.0.0
azure-mgmt-cosmosdb==0.14.0
azure-mgmt-datalake-analytics==0.2.1
azure-mgmt-datalake-nspkg==3.0.1
azure-mgmt-datalake-store==0.5.0
azure-mgmt-datamigration==0.1.0
azure-mgmt-deploymentmanager==0.2.0
azure-mgmt-devtestlabs==4.0.0
azure-mgmt-dns==2.1.0
azure-mgmt-eventgrid==2.2.0
azure-mgmt-eventhub==4.0.0
azure-mgmt-hdinsight==1.4.0
azure-mgmt-imagebuilder==0.4.0
azure-mgmt-iotcentral==3.0.0
azure-mgmt-iothub==0.12.0
azure-mgmt-iothubprovisioningservices==0.2.0
azure-mgmt-keyvault==2.2.0
azure-mgmt-kusto==0.3.0
azure-mgmt-loganalytics==0.6.0
azure-mgmt-managedservices==1.0.0
azure-mgmt-managementgroups==0.2.0
azure-mgmt-maps==0.1.0
azure-mgmt-marketplaceordering==0.2.1
azure-mgmt-media==2.2.0
azure-mgmt-monitor==0.9.0
azure-mgmt-msi==0.2.0
azure-mgmt-netapp==0.8.0
azure-mgmt-network==10.2.0
azure-mgmt-nspkg==3.0.2
azure-mgmt-policyinsights==0.4.0
azure-mgmt-privatedns==0.1.0
azure-mgmt-rdbms==2.2.0
azure-mgmt-recoveryservices==0.4.0
azure-mgmt-recoveryservicesbackup==0.6.0
azure-mgmt-redhatopenshift==0.1.0
azure-mgmt-redis==7.0.0rc1
azure-mgmt-relay==0.1.0
azure-mgmt-reservations==0.6.0
azure-mgmt-resource==10.0.0
azure-mgmt-search==2.1.0
azure-mgmt-security==0.4.1
azure-mgmt-servicebus==0.6.0
azure-mgmt-servicefabric==0.4.0
azure-mgmt-signalr==0.3.0
azure-mgmt-sql==0.18.0
azure-mgmt-sqlvirtualmachine==0.5.0
azure-mgmt-storage==11.0.0
azure-mgmt-trafficmanager==0.51.0
azure-mgmt-web==0.46.0
azure-multiapi-storage==0.3.5
azure-nspkg==3.0.2
azure-storage-blob==1.5.0
azure-storage-common==1.4.2
Babel==2.8.0
backcall==0.2.0
base58==2.0.1
bcrypt==3.1.7
beautifulsoup4==4.9.1
billiard==3.6.3.0
bleach==3.1.5
blinker==1.4
boto3==1.14.9
botocore==1.17.9
cached-property==1.5.1
cachetools==4.1.0
celery==4.4.5
celery-once==3.0.1
certifi==2020.6.20
cffi==1.14.0
chardet==3.0.4
click==7.1.2
colorama==0.4.3
cryptography==2.9.2
cycler==0.10.0
decorator==4.4.2
defusedxml==0.6.0
dnspython==1.16.0
docker==4.2.1
docutils==0.15.2
email-validator==1.1.1
entrypoints==0.3
enum-compat==0.0.3
et-xmlfile==1.0.1
fabric==2.5.0
flake8==3.8.3
flake8-polyfill==1.0.2
Flask==1.1.2
Flask-BabelEx==0.9.4
Flask-Login==0.5.0
Flask-Mail==0.9.1
Flask-Principal==0.4.0
Flask-Security-Too==3.4.3
Flask-Webpack==0.1.0
Flask-WTF==0.14.3
future==0.18.2
gast==0.3.3
gitdb==4.0.5
GitPython==3.1.3
google-auth==1.18.0
google-auth-oauthlib==0.4.1
google-pasta==0.2.0
grpcio==1.30.0
h5py==2.10.0
honcho==1.0.1
humanfriendly==4.18
idna==2.9
imageio==2.8.0
imgaug==0.4.0
inflection==0.5.0
invoke==1.4.1
ipykernel==5.3.0
ipython==7.15.0
ipython-genutils==0.2.0
ipywidgets==7.5.1
isodate==0.6.0
isort==4.3.21
itsdangerous==1.1.0
javaproperties==0.5.1
jdcal==1.4.1
jedi==0.17.1
Jinja2==2.11.2
jmespath==0.10.0
joblib==0.15.1
jsmin==2.2.2
jsondiff==1.2.0
jsonschema==3.2.0
jupyter==1.0.0
jupyter-client==6.1.3
jupyter-console==6.1.0
jupyter-core==4.6.3
Keras-Preprocessing==1.1.2
kiwisolver==1.2.0
knack==0.7.1
kombu==4.6.10
lazy-object-proxy==1.4.3
loggly-python-handler==1.0.1
lxml==4.5.1
Mako==1.1.3
mando==0.6.4
Markdown==3.2.2
MarkupSafe==1.1.1
matplotlib==3.2.2
mccabe==0.6.1
mistune==0.8.4
mock==2.0.0
more-itertools==8.4.0
msal==1.0.0
msal-extensions==0.1.3
msrest==0.6.16
msrestazure==0.6.3
mysql-connector-python==8.0.16
nbconvert==5.6.1
nbformat==5.0.7
networkx==2.4
notebook==6.0.3
numpy==1.19.0
oauthlib==3.1.0
opencv-python==4.2.0.34
opencv-python-headless==4.2.0.34
openpyxl==3.0.4
opt-einsum==3.2.1
packaging==20.4
pandas==1.0.5
pandocfilters==1.4.2
paramiko==2.7.1
parso==0.7.0
passlib==1.7.2
pathtools==0.1.2
pbr==5.4.5
pexpect==4.8.0
pickleshare==0.7.5
piexif==1.1.3
Pillow==7.1.2
pkginfo==1.5.0.1
pluggy==0.13.1
portalocker==1.7.0
prometheus-client==0.8.0
prompt-toolkit==3.0.5
protobuf==3.12.2
psutil==5.7.0
ptyprocess==0.6.0
py==1.8.2
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.6.0
pycparser==2.20
pydeck==0.4.0b2
pydocumentdb==2.3.5
pyflakes==2.2.0
Pygments==2.6.1
PyJWT==1.7.1
pylint==2.5.3
PyNaCl==1.4.0
pyOpenSSL==19.1.0
pyparsing==2.4.7
PyQRCode==1.2.1
pyrsistent==0.16.0
pytest==5.4.3
python-dateutil==2.8.1
python-editor==1.0.4
python-json-logger==0.1.11
pytz==2019.1
PyWavelets==1.1.1
PyYAML==5.3.1
pyzmq==19.0.1
qtconsole==4.7.5
QtPy==1.9.0
radon==4.1.0
redis==3.5.3
requests==2.24.0
requests-futures==1.0.0
requests-oauthlib==1.3.0
rsa==4.6
s3transfer==0.3.3
scikit-image==0.17.2
scikit-learn==0.23.1
scipy==1.4.1
scp==0.13.2
Send2Trash==1.5.0
Shapely==1.7.0
six==1.15.0
smmap==3.0.4
soupsieve==2.0.1
speaklater==1.3
SQLAlchemy==1.3.17
sshtunnel==0.1.5
streamlit==0.62.0
tabulate==0.8.7
tensorboard==2.2.2
tensorboard-plugin-wit==1.6.0.post3
tensorflow==2.2.0
tensorflow-estimator==2.2.0
termcolor==1.1.0
terminado==0.8.3
testpath==0.4.4
threadpoolctl==2.1.0
tifffile==2020.6.3
toml==0.10.1
toolz==0.10.0
tornado==6.0.4
tqdm==4.46.1
traitlets==4.3.3
tzlocal==2.1
urllib3==1.25.9
uWSGI @ file:///Users/runner/miniforge3/conda-bld/uwsgi_1589314754489/work
validators==0.15.0
vine==1.3.0
visual-logging==1.0
vsts==0.1.25
vsts-cd-manager==1.0.2
watchdog==0.10.2
wcwidth==0.2.5
webencodings==0.5.1
websocket-client==0.56.0
Werkzeug==1.0.1
widgetsnbextension==3.5.1
wrapt==1.12.1
WTForms==2.3.1
xenon==0.7.0
xlrd==1.2.0
xmltodict==0.12.0

```

```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
"""
Redis connection tests
"""

import time


def test_one(celery_worker):

    pass


def test_two():

    while True:
        print("sleep...")
        time.sleep(1)

```

</p>
</details>

# Expected Behavior

After `test_one` finishes, redis connections opened by `celery_worker` get closed.

# Actual Behavior

redis connections opened by `celery_worker` fixture don't get closed.

Output of `lsof -a -d 0-999` when `py.test` enters second test:

```
com.docke  4467 kuba  101u     IPv6 0xeaa391c3e1f5c49f          0t0                 TCP *:6379 (LISTEN)
com.docke  4467 kuba  110u     IPv6 0xeaa391c3decf7e7f          0t0                 TCP localhost:6379->localhost:59620 (ESTABLISHED)
com.docke  4467 kuba  112u     IPv6 0xeaa391c3decf5fdf          0t0                 TCP localhost:6379->localhost:59621 (ESTABLISHED)
python    44271 kuba    6u     IPv4 0xeaa391c3e683004f          0t0                 TCP localhost:59620->localhost:6379 (ESTABLISHED)
python    44271 kuba   11u     IPv4 0xeaa391c3e68388cf          0t0                 TCP localhost:59621->localhost:6379 (ESTABLISHED)
```

### PR DESCRIPTION
**Pull request by phanky1**
## Description

Fixes #6382.

The function-scoped fixtures (`celery_app/celery_worker`) build a new app for every test and never close its backend connections. Only the garbage collector releases them, and it runs too rarely to keep up so a long test run exhausts the open-file limit.

This PR forces a garbage collection on every app teardown.

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
can't we test the changes?

**Comment by phanky1** [comment]
> can't we test the changes?

I added regression tests.

**Comment by auvipy** [comment]
@Nusnus it would be great if you can check this once

**Comment by auvipy** [comment]
it it possible to additional tests?

**Comment by auvipy** [comment]
@phanky1 would you mind cross checking the last suggestion I merged? also explore the possibility of adding integration or smoke test for this change? since this is a long standing issue we need to be very cautious

**Comment by phanky1** [comment]
> @phanky1 would you mind cross checking the last suggestion I merged? also explore the possibility of adding integration or smoke test for this change? since this is a long standing issue we need to be very cautious

While the Copilot suggestion is valid, I think we shouldn't apply it because:
1. the leak it fixes rarely happens
2. it adds overhead to the whole test suite, because gc.collect() runs for every test which caused the last CI run to time out

I think in the worst case there would be just a few connection leaks during the run, so I decided to revert your change.
btw i added a new integration test.

As for the test failures now, there seems to be a GitHub issue, I will rerun the pipeline later.

**Comment by auvipy** [comment]
the unit tests are failing again!, can you please find out its root cause?

**Comment by phanky1** [comment]
> the unit tests are failing again!, can you please find out its root cause?

I see this error
>  × git clone --filter=blob:none --quiet https://github.com/celery/kombu.git /tmp/pip-req-build-tjzuhuk0 did not run successfully.
  │ exit code: 128
  ╰─> No available output.

Only the runs from blacksmith got this, I guess something is wrong with the access token there.

**Comment by auvipy** [comment]
i fixed them on main branch

**Review by auvipy — APPROVED** [review]

**Comment by auvipy** [comment]
thanks for your work here @phanky1 ! can you please investigate this issue https://github.com/celery/celery/issues/4511 ? this might be somewhat connected

**Comment by Nusnus** [comment]
@auvipy 
> @Nusnus it would be great if you can check this once

Thank you for flagging. The PR indeed introduces some anti-patterns.
Working on the fix right now.

P.S
Generally speaking, a call to `gc.collect()` is 99% a sign that something is wrong / this is not the right fix.

**Comment by auvipy** [comment]
OK

**Comment by auvipy** [comment]
@phanky1 can you please recheck?

**Comment by phanky1** [comment]
> @auvipy
> 
> > @Nusnus it would be great if you can check this once
> 
> Thank you for flagging. The PR indeed introduces some anti-patterns. Working on the fix right now.
> 
> P.S Generally speaking, a call to `gc.collect()` is 99% a sign that something is wrong / this is not the right fix.

Thanks for bringing it up. After further consideration, I think we can just remove that gc.collect() call and let the garbage collector do its job naturally. I think it would take a big test suite for this issue to become that alarming, so GC should have plenty of chances to run and clean things up to keep the number of orphaned backends low.

I have this PR https://github.com/celery/celery/pull/10648.
